### PR TITLE
Only provide fallback options for host analyzers

### PR DESCRIPTION
**Pull request by sharwell**
* Separate analysis passes into "project" analyzers (aka NuGet-installed analyzers) and "solution" analyzers (aka host- or VSIX-installed analyzers)
* Only provide fallback options from the host for host-installed analyzers

### Items not covered by this pull request

* Redirect Code Style analyzers to Features analyzers when loaded in Visual Studio. This is addressed separately by #75250.
* Treat Features analyzers as project analyzers when they are being used in place of Code Style analyzers.
* NuGet-installed code fixes and refactorings should not use host options, but currently they do. Only source generators and analyzers are corrected by this pull request.

### Known issues introduced by this change

* Prior to this change, NuGet-installed diagnostic suppressors were able to suppress diagnostics reported by VSIX-installed analyzers. This is no longer possible. To resolve this issue, NuGet-installed analyzers need to be added to _both_ `CompilationWithAnalyzers` instances, but for the host instance, only the diagnostic suppressors should run. #75399

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Diagnostics/DiagnosticAnalyzerDriver/DiagnosticAnalyzerDriverTests.cs
Comment (tmat) [inline src/EditorFeatures/CSharpTest/Diagnostics/DiagnosticAnalyzerDriver/DiagnosticAnalyzerDriverTests.cs:660]: Link issue

File: src/EditorFeatures/Core/EditorConfigSettings/DataProvider/SettingsProviderBase.cs
Comment (tmat) [inline src/EditorFeatures/Core/EditorConfigSettings/DataProvider/SettingsProviderBase.cs:115]: Perhaps we can store the value of `ConfigOptionsWithoutFallback` and `_projectDirectoryConfigData.Value` in fields instead of `AnalyzerConfigData`, so that we don't need to qualify them everywhere.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs:310]: Keep `AnalyzerOptions` and rename `HostAnalyzerOptions` to `AnalyzerOptionsWithFallback`?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.RegularCompilationTracker_Generators.cs
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.RegularCompilationTracker_Generators.cs:399]: Source generators shouldn't use fallback?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs
Comment (tmat) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs:228]: I'd keep the name as is (GetAnalyzerConfigOptionsAsync).

Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs:228]: Especially since this is actually not using the fallback -- Project.AnalyzerOptions are not falling back anymore:

public AnalyzerOptions AnalyzerOptions => State.ProjectAnalyzerOptions;

File: src/Workspaces/Core/Portable/Workspace/Solution/Project.cs
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:155]: AnalyzerOptionsWithFallback?

File: src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs
Comment (tmat) [inline src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs:579]: Perhaps the body of the loop can be a local function and AppendAnalyzerMap can be inlined.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs
Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs:228]: I'm assuming it's `GetHostAnalyzerConfigOptionsAsync` so that we don't need to `ifdef` the call sites,
but I'd prefer the non-extension method `Document.GetAnalyzerConfigOptionsAsync` that falls back to also not be renamed (not include "Host").

File: src/EditorFeatures/CSharpTest/Diagnostics/DiagnosticAnalyzerDriver/DiagnosticAnalyzerDriverTests.cs
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Diagnostics/DiagnosticAnalyzerDriver/DiagnosticAnalyzerDriverTests.cs:660]: I will create an issue to link if I don't find a clean way to fix this before merging.

File: src/EditorFeatures/Core/EditorConfigSettings/DataProvider/SettingsProviderBase.cs
Comment (sharwell) [reply src/EditorFeatures/Core/EditorConfigSettings/DataProvider/SettingsProviderBase.cs:115]: I'm OK with being overly explicit in this PR. It was part of the design for reducing risk in it.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.RegularCompilationTracker_Generators.cs
Comment (sharwell) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.RegularCompilationTracker_Generators.cs:399]: No, there is no such thing as a host-installed source generator. All source generators in all cases must ignore host options.

File: src/Workspaces/Core/Portable/Workspace/Solution/Project.cs
Comment (sharwell) [reply src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:155]: ➡️ I find `Host` to be more meaningful than `WithFallback` in all cases except for the actual fallback implementation.

**Comment by arkalyanms** [comment]
What do we need before this can be published?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs:228]: ➡️ This is under the umbrella of code fixes and refactorings still use host options even though they shouldn't. Or, in the case of this conditional invocation, it uses a mix of options. It's not ideal, but I'll leave it for future work.

File: src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs
Comment (sharwell) [reply src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs:579]: ➡️ This is small enough that I'll leave it for now.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs:230]: ➡️ Comment added

File: src/EditorFeatures/CSharpTest/Diagnostics/DiagnosticAnalyzerDriver/DiagnosticAnalyzerDriverTests.cs
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Diagnostics/DiagnosticAnalyzerDriver/DiagnosticAnalyzerDriverTests.cs:660]: ➡️ Link added to #75399

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs:364]: Deduplicate code in a follow up. E.g. use a common base class

**Review by tmat — APPROVED** [review]
Approved assuming a clean-up will be done in a follow up PR.

**Pull request merged**
